feat: chain sync projection primitives

mkazlauskas opened the pull request:
# Context

In order to create a database that is optimized for our provider queries, we'll need to project events from Chain Sync protocol 

# Proposed Solution

See [README](https://github.com/input-output-hk/cardano-js-sdk/blob/feat/generalized-block-handler/packages/projection/README.md) of the new `projection` package and the last commit (cc4bcea578a8da4b29d9fc14869f00a2090bcd70)

# Important Changes Introduced

See the commit log.

## Follow-up work

`projection`:
- Create an abstraction for "store projectors" and implementation for stake pools and stake keys
- Implement postgres store/repository for stake pools and stake keys

`ogmios`:
- Complete the Ogmios transaction to core mapping. This PR only added mapping of certificates, the rest of transaction fields are currently not mapped.
- Create a source observable from Ogmios ChainSync client. Also, create an abstraction to initialize `evtCache$` for [withRolledBackEvents](https://github.com/input-output-hk/cardano-js-sdk/blob/feat/generalized-block-handler/packages/projection/src/operators/withRolledBackEvents.ts#L27)

`golden-test-generator`:
- Write tests for existing functionality
- Start Ogmios ChainSync from some block (not necessarily from origin)

rhyslbw commented on packages/core/src/Provider/Provider.ts:
```diff
@@ -1,4 +1,5 @@
 import { Percent } from '../Cardano';
+// eslint-disable-next-line import/no-extraneous-dependencies
```
This is curious. Seems like this lint rule is a bit naive

mkazlauskas replied:
Yeah, it doesn't know that it's just a type. If we start running into more of those we could loosen it up by allowing devDependencies in the rule

rhyslbw replied:
Next update might resolve this: https://github.com/import-js/eslint-plugin-import/blob/main/CHANGELOG.md#unreleased

rhyslbw commented on packages/projection/package.json:
```diff
@@ -0,0 +1,75 @@
+{
+  "name": "@cardano-sdk/projection",
+  "version": "0.0.1",
```
```suggestion
  "version": "0.1.0",
```

Conventionally we've been starting at this version

mkazlauskas replied:
Updated to 0.1.0 (sorry should've made a fixup for this, but already did it when squashing the util-rxjs type util commit

rhyslbw commented on packages/projection/src/operators/withRolledBackEvents.ts:
```diff
@@ -0,0 +1,81 @@
+                // Review: consider checking if `eventCache` isn't missing some blocks due to
+                // not syncing from genesis AND not providing sufficient evtCache$.
+                // Could be nice to have, but also could be redundant
+                // if we test our implementation of creating evtCache$
```
Maybe just add this then?

mkazlauskas replied:
Done in 03cfe9408d1f2e37273a33c921e91e2c3e6b9455

rhyslbw reviewed (changes requested):
Looks really nice so far @mkazlauskas!

1. 8a41a2fdb899f659be76d1ccf219832396c7025f seems like a `fix`.
2. Let's drop cae91f0c3c87a669033213e1897216c74b9afcba given there's no implementation to give us test coverage.
3. I think we should be keeping the `projection` package free from Ogmios implementations, as it makes sense for those to be scoped to the `ogmios` package.

mkazlauskas commented on yarn-project.nix:
```diff
@@ -180,15 +180,11 @@ cacheEntries = {
 "interpret@npm:1.4.0" = { filename = "interpret-npm-1.4.0-17b4b5b0a4-2e5f51268b.zip"; sha512 = "2e5f51268b5941e4a17e4ef0575bc91ed0ab5f8515e3cf77486f7c14d13f3010df9c0959f37063dcc96e78d12dc6b0bb1b9e111cdfe69771f4656d2993d36155"; };
 "bs-logger@npm:0.2.6" = { filename = "bs-logger-npm-0.2.6-7670f88b66-d34bdaf68c.zip"; sha512 = "d34bdaf68c64bd099ab97c3ea608c9ae7d3f5faa1178b3f3f345acd94e852e608b2d4f9103fb2e503f5e69780e98293df41691b84be909b41cf5045374d54606"; };
 "semver@npm:7.3.8" = { filename = "semver-npm-7.3.8-25a996cb4f-ba9c7cbbf2.zip"; sha512 = "ba9c7cbbf2b7884696523450a61fee1a09930d888b7a8d7579025ad93d459b2d1949ee5bbfeb188b2be5f4ac163544c5e98491ad6152df34154feebc2cc337c1"; };
-<<<<<<< HEAD
```
Nice catch, fixed it

rhyslbw replied:
Still has a conflict 
https://github.com/input-output-hk/cardano-js-sdk/pull/520/files#diff-6d359c8592cd08a5443a4ac76549b2283066c245d8b9d84e73ecb38414b4d660R183

mkazlauskas replied:
It was actually another conflict :man_facepalming: There must be something wrong with my rebase workflow, I'll figure it out next time I need to merge lock files when rebasing

mkazlauskas commented:
Addressed requested changes and added 1 of my own: b548aece1d2c5ce99faaf99aee2850332126ecbd replaced `blockHeight` with `slot` in `CertificatePointer` because this type is very similar to the pointer used in 'pointer addresses', and it uses slots. Better keep them aligned.

Ready for re-review.

mirceahasegan commented on packages/golden-test-generator/src/ChainSync/chainSync.ts:
```diff
@@ -0,0 +1,126 @@
+
+type RequestedBlocks = { [blockHeight: number]: Ogmios.Schema.Block };
+
+const getBlockHeaderAndHash = (block: Ogmios.Schema.Block) => {
```
You could use `ogmiosToCore.getBlock` utility to get the required info without having to handle specific block types:
```diff
+ const {header: {blockNo: blockHeight, hash, slot}} = ogmiosToCore.getBlock(block);

mkazlauskas replied:
Took it a step further and updated `golden-test-generator` chain-sync to write blocks in `core` format instead of Ogmios:
e6932f013ca5a1b98b4c3b60a6a10145f51c22d5
8a26d22ba796cf8302fd34109b13306664331848
580e09e8e1c257d184c8afdab85fe82315145746
5e5881e9aa505a45d5f6d58c5296b7240755dc55
4524a62a7dcdf12ea2e5b5b119a1b3662ffaf350

rhyslbw replied:
580e09e8e1c257d184c8afdab85fe82315145746 should be included with the changes in 4524a62a7dcdf12ea2e5b5b119a1b3662ffaf350. Might need to do some squashing to have clarity over the atomic changes and the appropriate sequence

mirceahasegan commented on packages/util-rxjs/test/blockingWithLatestFrom.test.ts:
```diff
@@ -0,0 +1,36 @@
+describe('blockingWithLatestFrom', () => {
+  it('waits for dependency$ to emit, does not lose any source emissions and subscribes only once', () => {
+    createTestScheduler().run(({ hot, expectObservable, expectSubscriptions }) => {
+      const source$ =     hot('a--b-c');
```
What is the expected behaviour when `source$` emits multiple values before `dependency$` emits?
Maybe add a test for it too.

mkazlauskas replied:
It will defer any emissions until `dependency$` emits and then emit all of them combined with the dependency

mkazlauskas replied:
Added some comments and updated the test in 797159e30af9e06843643d7bb593f9f9c9d7d35d

mirceahasegan reviewed (dismissed):
LGTM 👽 . I would prefer a more in-depth re-review after the squash

rhyslbw commented on packages/golden-test-generator/src/index.ts:
```diff
@@ -70,47 +70,67 @@ program
+      });
+
+program
+  .command('chain-sync')
```
I'm not sure this is a better name for the command as it's not describing the outcome, and chain-sync is just the underlying protocol. How about `project-chain-to-file` or something like that?

mkazlauskas replied:
How about `chain-sync-events`?

rhyslbw replied:
Is the output event data or block data?

mkazlauskas replied:
Event data: either roll forward or roll backward

rhyslbw replied:
Let's go with `chain-sync-events`

mkazlauskas replied:
Done in 6e460e50326939386ed249eef18ac9735d04f1a7. It will create a conflict with 42e7c3d2230fe54f5e6bee67dd5027585f35c67f when rebasing, but I think it should be ok to just squash the 3 commits into 1.

rhyslbw reviewed (changes requested):
Now would be a good time to remove the [ignore for the GTG package](https://github.com/input-output-hk/cardano-js-sdk/blob/f1cebb4d795847e3f992d1f6869dcac47842d5e6/.eslintignore#L4) if it's able to be made compliant without to much effort.

mkazlauskas commented on packages/projection/README.md:
```diff
@@ -0,0 +1,20 @@
+# Cardano JS SDK | projection
+
+Chain Sync event projection utilities.
```
It's [coupled](https://github.com/input-output-hk/cardano-js-sdk/pull/520/files#diff-91c2be13a76d158ec0354939267e38fcc1e9101449841d2b62f1df2729eb65b6R1) specifically with chain-sync though

rhyslbw replied:
Let's leave as-is for now.

rhyslbw commented on packages/core/src/CardanoNode/types/CardanoNode.ts:
```diff
@@ -68,3 +68,31 @@ export interface CardanoNode {
+  RollBackward
+}
+
+export interface ChainSyncRollForward {
```
After offline discussion it was determined to be considered in a future refactor as-needed.

mkazlauskas commented on packages/wallet/package.json:
```diff
@@ -53,7 +53,7 @@
   "devDependencies": {
     "@cardano-foundation/ledgerjs-hw-app-cardano": "^5.0.0",
-    "@cardano-ogmios/client": "5.5.5",
+    "@cardano-ogmios/client": "5.5.7",
```
@Rhys was this change in an inappropriate commit? Clicking on the file here takes me to
> We went looking everywhere, but couldn’t find those commits.

:+1: for `including the Docker Compose files.`

mkazlauskas replied:
Updated docker-compose in b5d0b74ba9664a196a504c9797f954be5ec38078 and verified that tag exists

rhyslbw replied:
> @Rhys was this change in an inappropriate commit?

Ah no, all good. I had two tabs open and thought I was reviewing a single commit. I can see it's part of 250b15b62a6c881d95e4f482418136f5684565f3

mkazlauskas commented on packages/projection/package.json:
```diff
@@ -0,0 +1,76 @@
+  "publishConfig": {
+    "access": "public"
+  },
+  "contributors": [
```
60d9ae7cd62b5c1ea800f3157692f16183775f9b

rhyslbw reviewed (approved):
Stellar work @mkazlauskas :rocket:

mirceahasegan reviewed (approved):
Great work! 🚀
